design/230728_tinkerbell_workflow_execution.md

```diff
@@ -0,0 +1,491 @@
+| `WorkflowRejected`  | Tink Agent has explicitly rejected the request to run the Workflow |
+
+### Cancelling a Workflow
+
+Operators can cancel Workflows by deleting the Workflow or setting the status to Cancelling.
+
+When Kubernetes resources are deleted they are removed from the system via a garbage collection process. Resources that specify [finalizers](https://kubernetes.io/docs/concepts/overview/working-with-objects/finalizers/) are are not garbage collected until all finalizers are removed.
+
+Tink Controller will populate a finalizer when it prepares a Workflow. The finalizer will remain on the object until it transitions to an end state.
+
+When a Workflow is deleted, Tink Controller will transition the Workflow to either Canceled (if the Workflow is in a Pending state) or Cancelling (if the Workflow is Scheduled or Running).
+
+### Handling a rejected Workflow
+
+When a Workflow is dispatched to Tink Agent it transitions to a Scheduled state. If Tink Agent rejects a Workflow, because its already running a Workflow, it will be transitioned back to Pending by Tink Server. The Workflow will not be scheduled again until a configurable time period has elapsed. The time period will grow with each retry attempt up-to a maximum.
+
+## Failure Modes
+
+### Supervising Workflows
+
+In addition to preparing Workflows, Tink Controller supervises Workflow execution. The core concerns Tink Controller supervises include:
+1. Ensuring Scheduled Workflows transition to a Running state within a finite time period. When the period has elapsed Tink Cotnroller will transition the Workflow to a Failed state.
+2. Ensuring Cancelling Workflows transition to a Canceled state within a finite time period. When the period has elapsed Tink Controller will transition the Workflow to a Canceled state with appropriate failure reason and message.
+3. Ensuring Running Workflows complete within the configured timeout.
+4. Ensuring individual Workflow Actions execute within the configured timeout.
+
+### What happens if 2 Tink Agents connect to Tink Server with the same identifier?
+
+When Tink Agents connect to Tink Server they will present uniquely identifying information. If 2 Tink Agent instances present the same identifying information the behavior is undefined.
+
+### What happens if 2 Workflows are dispatched to Tink Agent (double dispatch)?
+
+Tink Server is responsible for controlling the flow of Workflows to Tink Agent. If Tink Server didn't have a record for an executing Workflow, or if an executing Workflows state was misrepresented on the Workflow resource, Tink Server may perform a double Workflow dispatch.
+
+If Tink Worker receives a second Workflow while currently executing a Workflow it will dispatch a Workflow rejection event.
+
+On receiving the Workflow rejection event Tink Server will transition the Workflow to a Pending state. The Workflow will not be scheduled again until a configurable period has elapsed.
```

Comment:
im assuming this is in reference to the 2nd workflow not the original and running workflow, correct? if so maybe some clarifying in this sentence would help.